client/protocols/hinting.py

```diff
@@ -0,0 +1,185 @@
+"""Implements the HintProtocol, which generates hints for students
+that are stuck on a coding question. The protocol uses analytics
+to determine whether a hint should be given and then
+obtains them from the hint generation server. Free response questions
+are posed before and after hints are provided.
+"""
+
+from client.sources.common import core
+from client.sources.common import models as sources_models
+from client.protocols.common import models as protocol_models
+from client.utils import auth
+from client.utils import format
+
+import json
+import logging
+import os
+import pickle
+import random
+import re
+import urllib.error
+import urllib.request
+
+log = logging.getLogger(__name__)
+
+#####################
+# Hinting Mechanism #
+#####################
+
+class HintingProtocol(protocol_models.Protocol):
+    """A protocol that provides rubber duck debugging and hints if applicable.
+    """
+
+    HINT_SERVER = "http://146.148.59.194:5000/"
+    HINT_ENDPOINT = 'api/hints'
+
+    def run(self, messages):
+        """Determine if a student is elgible to recieve a hint. Based on their
+        state, poses reflection questions.
+
+        After more attempts, ask if students would like hints. If so, query
+        the server.
+        """
+        if self.args.local:
+            return
+
+        if 'analytics' not in messages:
+            log.info('Analytics Protocol is required for hint generation')
+            return
+        if 'file_contents' not in messages:
+            log.info('File Contents needed to generate hints')
+            return
+
+        messages['hinting'] = {}
+        history = messages['analytics']['history']
+        questions = history['questions']
+        current_q = history['question']
+
+        for question in current_q:
+            if question not in questions:
+                continue
+            stats = questions[question]
+            messages['hinting'][question] = {'prompts': {}, 'reflection': {}}
+            hint_info = messages['hinting'][question]
+
+            # Only prompt
+            if (stats['solved'] and stats['attempts'] > 6):
+                hint_info['elgible'] = False
+                if self.args.question:
+                    # Only prompt for reflection with question specified.
+                    log.info('Giving reflection response on %s', question)
+                    reflection = random.choice(SOLVE_SUCCESS_MSG)
+                    if not confirm("Nice work! Could you answer a quick question"
+                                   " about how you approached this question?"):
+                        hint_info['reflection']['accept'] = False
+                    else:
+                        hint_info['reflection']['accept'] = True
+                    prompt_user(reflection, hint_info)
```

Comment:
Just to clarify, are you prompting the user even if they refuse to answer a quick question?